File: java/src/dsclient/scheduling_policies/BestFit.java
package dsclient.scheduling_policies;

import java.io.IOException;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import dsclient.JobDispatchPolicy;
import dsclient.JobScheduler;
import dsclient.models.JobSubmission;
import dsclient.models.Resource;

public class BestFit implements JobDispatchPolicy {
    public void dispatch(JobScheduler scheduler) throws IOException {
        while (true) {
            String received = scheduler.inquire("REDY");
            if (received.equals("NONE")) {
                break;
            }

            JobSubmission job = JobSubmission.fromReceivedLine(received);

            List<Resource> resources = scheduler.fetchCapableResources(job);
            List<Resource> sufficient = resources.stream().filter(
                res -> res.coreCount >= job.coreCount
                && res.disk >= job.disk
                && res.memory >= job.memory
            ).collect(Collectors.toList());

            Resource best;
            if (sufficient.isEmpty()) {
                best = resources.get(0);
            } else {
                best = sufficient.stream().min(
                    Comparator.<Resource>comparingInt(res -> res.coreCount - job.coreCount)
                            .thenComparingInt(res -> res.availableTime)
                ).get();
            }

            scheduler.schedule(job, best);
        }
    }
}

File: python/ds_client/scheduling_policies/best_fit.py

from __future__ import annotations
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from ..job_scheduler import JobScheduler

from ..job_dispatch_policy import JobDispatchPolicy
from ..models import JobSubmission

class BestFit(JobDispatchPolicy):
    def dispatch(self, scheduler: JobScheduler) -> None:
        while True:
            received = scheduler.inquire('REDY')
            if received == 'NONE':
                break

            job = JobSubmission.from_received_line(received)

            resources = scheduler.fetch_capable_resources(job)
            sufficient = [
                res for res in resources
                if (
                    res.core_count >= job.core_count
                    and res.disk >= job.disk
                    and res.memory >= job.memory
                )
            ]

            best = None
            if sufficient:
                best = min(
                    sufficient,
                    key=lambda res: (
                        res.core_count - job.core_count,
                        res.available_time,
                    ),
                )
            else:
                best = resources[0]

            scheduler.schedule(job, best)
